internal/test/e2e/s3client.go

```diff
@@ -0,0 +1,430 @@
+package e2e
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	s3aws "github.com/aws/aws-sdk-go/service/s3"
+)
+
+type (
+	s3TestClient struct {
+		s3 *s3aws.S3
+	}
+)
+
+// request and response types
+type (
+	bucketInfo struct {
+		creationDate time.Time
+		name         string
+	}
+
+	copyObjectResponse struct {
+		lastModified time.Time
+		etag         string
+	}
+
+	listBucketResponse struct {
+		buckets []bucketInfo
+	}
+
+	listObjectsOptions struct {
+		prefix    string
+		marker    string
+		delimiter string
+		maxKeys   int64
+	}
+
+	listObjectsResponse struct {
+		contents       []headObjectResponse
+		commonPrefixes []string
+		nextMarker     string
+		truncated      bool
+	}
+
+	completePart struct {
+		partNumber int64
+		etag       string
+	}
+
+	getObjectOptions struct {
+		offset int64
+		length int64
+	}
+
+	getObjectResponse struct {
+		body         io.ReadCloser
+		etag         string
+		lastModified time.Time
+		metadata     map[string]string
+	}
+
+	headObjectResponse struct {
+		contentLength int64
+		etag          string
+		key           string
+		lastModified  time.Time
+		metadata      map[string]string
+	}
+
+	multipartUploadInfo struct {
+		bucket   string
+		key      string
+		uploadID string
+	}
+
+	putObjectOptions struct {
+		metadata map[string]string
+	}
+
+	putObjectPartOptions struct {
+	}
+
+	putObjectResponse struct {
+		etag string
+	}
+
+	putObjectPartResponse struct {
+		etag string
+	}
+
+	uploadInfo struct {
+		bucket   string
+		etag     string
+		key      string
+		uploadID string
+	}
+)
+
+func (c *s3TestClient) Config() aws.Config {
+	return c.s3.Config
+}
+
+func (c *s3TestClient) AbortMultipartUpload(bucket, key string, uploadID string) error {
+	var input s3aws.AbortMultipartUploadInput
+	input.SetBucket(bucket)
+	input.SetKey(key)
+	input.SetUploadId(uploadID)
+	_, err := c.s3.AbortMultipartUpload(&input)
+	return err
+}
+
+func (c *s3TestClient) CompleteMultipartUpload(bucket, object, uploadID string, parts []completePart, opts putObjectOptions) (uploadInfo, error) {
+	var input s3aws.CompleteMultipartUploadInput
+	input.SetBucket(bucket)
+	input.SetKey(object)
+	input.SetUploadId(uploadID)
+	var upload s3aws.CompletedMultipartUpload
+	var inputParts []*s3aws.CompletedPart
+	for i := range parts {
+		inputParts = append(inputParts, &s3aws.CompletedPart{
+			PartNumber: &parts[i].partNumber,
+		})
+		upload.SetParts(inputParts)
+	}
+	input.SetMultipartUpload(&upload)
+
+	resp, err := c.s3.CompleteMultipartUpload(&input)
+	if err != nil {
+		return uploadInfo{}, err
+	}
+	return uploadInfo{
+		bucket:   *resp.Bucket,
+		etag:     *resp.ETag,
```

Comment:
Is it possible we are dereferencing `nil` here? iirc etag and one other field are omitempty

Reply:
I don't think you can create an object through the S3 api that doesn't have an etag since the worker will create and set it implicitly. So for the sake of not adding too much verbosity and because it's just a test client, I have been very aggressively dereferencing without checking anything.